Modify the SSL certification assignment and cleanup

Tusamarco opened the pull request:
Donor doesn't need any clone_ssl but it uses the ssl_% server section While joiner requires the clone_ssl_%

Tusamarco committed:
Modify the SSL certification assignment and cleanup
Donor doesn't need any clone_ssl but it uses the ssl_% server section
While joiner requires the clone_ssl_%

Tusamarco commented:
Still working on the other fixes

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,85 +320,157 @@ setup_clone_plugin()
+    CLIENT_SSL_MODE=$(parse_cnf sst ssl_mode "")
 
-    if [ -z "$CLIENT_SSL_CERT" -o "$CLIENT_SSL_CERT" == "" ]
+    if [ -z "$CLIENT_SSL_CERT" ] || [ "$CLIENT_SSL_CERT" == "" ]
```
`if [ -z "$CLIENT_SSL_CERT" ]` is enough

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,85 +320,157 @@ setup_clone_plugin()
+                $MYSQL_ACLIENT -e "SET GLOBAL clone_ssl_key='$CLONE_SSL_KEY'"
+                $MYSQL_ACLIENT -e "SET GLOBAL clone_ssl_ca='$CLONE_SSL_CA'"
+                if [ "$CLEANUP_CLONE_PLUGIN" == "yes" ]; then
+                    CLEANUP_CLONE_SSL="yes"
```
`CLEANUP_CLONE_SSL` is unused variable now

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,85 +320,157 @@ setup_clone_plugin()
 
-    wsrep_log_debug "-> CLONE_SSL_CERT: $CLONE_SSL_CERT; CLONE_SSL_KEY: $CLONE_SSL_KEY; CLONE_SSL_CA: $CLONE_SSL_CA"
+    # Once identify and configured the CLONE plugin we need to check if the DONOR and the 
+    # Joiner handle encrypted connctions.
```
s/connctions/connections/

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,85 +320,157 @@ setup_clone_plugin()
     fi
 
-    wsrep_log_debug "-> CLONE_SSL_CERT: $CLONE_SSL_CERT; CLONE_SSL_KEY: $CLONE_SSL_KEY; CLONE_SSL_CA: $CLONE_SSL_CA"
+    # Once identify and configured the CLONE plugin we need to check if the DONOR and the 
```
spaces at the end of lines

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,85 +320,157 @@ setup_clone_plugin()
+    SERVER_SSL_VALID="yes"
+
+    # If pxc_encrypt_cluster_traffic is set ON then we go through the ssl checks
+    # otherwise we skip and do not set SSL    
```
spaces at the EOL

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,85 +320,157 @@ setup_clone_plugin()
     then
-        wsrep_log_info "CLONE SSL not configured. Checking general MySQL SSL settings."
+        # Client information must check on both side to allow connection from Donor to Joiner
+        wsrep_log_debug "-> CLONE_SSL_CERT: $CLONE_SSL_CERT; \ 
```
space at the EOL. \ is preceding the space, so does not take effect

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,85 +320,157 @@ setup_clone_plugin()
+            # We actually modify the variables only if we are on the Joiner
+            if [ -n "$CLONE_SSL_CERT" ] && [ -n "$CLONE_SSL_KEY" ] && [ "$ROLE" == "joiner" ]
+            then
+                wsrep_log_info "Using SSL configuration from MySQL Server."
```
Let's say donor has clone plugin installed and set server certificates. `pxc_encrypt_cluster_traffic=off`. This means:
1. cloning from donor will require clone client to use secure connection
2. galera communication channel is not encrypted
Am I right?

Havig the above logic on Joiner side, clone plugin will be installed, but `clone_ssl_*` will not be set.

In my opinion the logic on joiner node should be something like:
1. Check if clone plugin installed. If not -> install it
2. Check if clone_ssl_ certificates set. If set -> leave as it is
3. if not set -> use certificates from [sst] section, but if pxc_encrypt_cluster_traffic=on use certificates from [mysqld] section if available

Tusamarco committed:
Merge branch 'PXC-4469-percona' into PXC-4469-fix-clonesst

* PXC-4469-percona:
  PXC-4573: SST Role lacks INNODB_REDO_LOG_ARCHIVE
  PXC-4498: PXC node "stalls" and causes the cluster to stop responding
  clang-tidy check fixed. 1. Boost url updated 2. 8.4 branch included

Tusamarco committed:
Modify the SSL certification assignment and cleanup
Donor doesn't need any clone_ssl but it uses the ssl_% server section
While joiner requires the clone_ssl_%
Refactoring after discussion with Kamil

Tusamarco committed:
too fast in committing the changes

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,98 +299,160 @@ setup_clone_plugin()
         if [ "$ROLE" == "donor" ]; then
```
We've got exactly the same `if` a few lines above, so we can set `CLEANUP_CLONE_PLUGIN="yes"` there

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,98 +299,160 @@ setup_clone_plugin()
+    CLIENT_SSL_VALID="no"
+    SERVER_SSL_VALID="no"
+
+    # Checking CLient certificates
```
s/CLient/client/

Tusamarco replied:
done

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,98 +299,160 @@ setup_clone_plugin()
-        fi
+         fi
+        # We set that client certificates are valid
+        CLIENT_SSL_VALID="yes"
```
As we spoke today, this whole part should happen only on joiner. It is about setting certificates on clone instance to be able to connect to donor instance. It has nothing to client SSL certificates validity. clone_ssl_cert plays any role only on Joiner side.

Client certificates are used ONLY on Donor side, when Donor connects to Joiner clone instance and triggers cloning. So it is not valid to say that CLIENT_SSL_VALID=true if CLONE_SSL_CERT is set. These are not related variables.

Tusamarco replied:
done

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,98 +299,160 @@ setup_clone_plugin()
+    # Checking Server certificates
+    # Check that we have all the files
+    # If they have not been explicitly specified, check the datadir
+    if [ -z "$SERVER_SSL_CA" ]; then
```
As agreed, let's remove these 3 ifs

Tusamarco replied:
done

kamil-holubicki commented on scripts/wsrep_sst_clone.sh:
```diff
@@ -328,98 +299,160 @@ setup_clone_plugin()
     CLIENT_SSL_OPTIONS=""
-    if [ -n "$CLONE_SSL_CERT" -a -n "$CLONE_SSL_KEY" ]
+    wsrep_log_debug ":->(POST) CLIENT_SSL_VALID=$CLIENT_SSL_VALID SERVER_SSL_VALID=$SERVER_SSL_VALID"
+    if [ "$CLIENT_SSL_VALID" == "yes" ] && [ "$SERVER_SSL_VALID" == "yes" ]
```
As we discussed, everything what happens here, affects only donor, so should happen only on donor (the result is setting 
1. `CLIENT_SSL_OPTIONS` - used during Donor -> clone instance connection
2. `REQUIRE_SSL` - creation of clone user on Donro server.
These are two independent things

Tusamarco replied:
done

Tusamarco committed:
Cleanup for SSL configuration splitting the assignments by role Donor/Joiner

Tusamarco committed:
Cleanup for SSL configuration splitting the assignments by role Donor/Joiner Part 2 (with more isolation in the assignments)

Tusamarco committed:
space

Tusamarco committed:
space and debug
